### ISSUE TITLE
Build failure: isort on Python 3.12

### ISSUE THREAD

**Issue by PerchunPak** [NixOS/nixpkgs#296222]
### Steps To Reproduce

Run `nix-shell -p python312Packages.isort` on [unstable channel](https://github.com/NixOS/nixpkgs/commit/d691274a972b3165335d261cc4671335f5c67de9).

### Build log

```
error: builder for '/nix/store/05908nd888ss63axnky4xv17swjvq2jd-python3.12-pydocstyle-6.3.0.drv' failed with exit code 1;
       last 10 log lines:
       >         assert function.name == 'do_something'
       >         assert function.decorators == []
       >         assert function.children == []
       > >       assert function.docstring == 'f"""Do some {foo} and some {bar}."""'
       > E       assert None == 'f"""Do some {foo} and some {bar}."""'
       > E        +  where None = Function(name='do_something', _source=['foo = "bar"\n', 'bar = "baz"\n', 'def do_something(pos_param0, pos_param1, kw_param0="default"):\n', '    f"""Do some {foo} and some {bar}."""\n', '    return None\n'], start=3, end=5, decorators=[], docstring=None, children=[], callable_args=['pos_param0', 'pos_param1', 'kw_param0'], parent=Module(name='file_path', _source=['foo = "bar"\n', 'bar = "baz"\n', 'def do_something(pos_param0, pos_param1, kw_param0="default"):\n', '    f"""Do some {foo} and some {bar}."""\n', '    return None\n'], start=1, end=6, decorators=[], docstring=None, children=[Function(name='do_something', _source=['foo = "bar"\n', 'bar = "baz"\n', 'def do_something(pos_param0, pos_param1, kw_param0="default"):\n', '    f"""Do some {foo} and some {bar}."""\n', '    return None\n'], start=3, end=5, decorators=[], docstring=None, children=[], callable_args=['pos_param0', 'pos_param1', 'kw_param0'], parent=Module(name='file_path', _source=['foo = "bar"\n', 'bar = "baz"\n', 'def do_something(pos_param0, pos_param1, kw_param0="default"):\n', '    f"""Do some {foo} and some {bar}."""\n', '    return None\n'], start=1, end=6, decorators=[], docstring=None, children=[...], parent=None, _dunder_all=None, dunder_all_error=None, future_imports=set(), skipped_error_codes=''), skipped_error_codes='')], parent=None, _dunder_all=None, dunder_all_error=None, future_imports=set(), skipped_error_codes=''), skipped_error_codes='').docstring
       >
       > src/tests/parser_test.py:100: AssertionError
       > ======================== 2 failed, 106 passed in 0.42s =========================
       > /nix/store/v099hqvw5z87423p4hz1vfhzaqa07dii-stdenv-linux/setup: line 1578: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/05908nd888ss63axnky4xv17swjvq2jd-python3.12-pydocstyle-6.3.0.drv'.
error: 1 dependencies of derivation '/nix/store/j1w7d8pp4pd294amwsz5nyv8hnj9ynvj-python3.12-pylama-8.4.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/pxf2yr5vr16vxns2f8phx1l4y8fmrmmj-python3.12-isort-5.12.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/33jgsahk9mq6zzww4shj1gicay9ifypx-home-manager-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/d4xg4yqwwb67cvaxc2azwghmyr09sxa1-man-paths.drv' failed to build
error: 1 dependencies of derivation '/nix/store/myzzls0cb0xmacmfsvkhb9311bbcjvfm-python3.12-isort-5.12.0-fish-completions.drv' failed to build
error: 1 dependencies of derivation '/nix/store/28s48xv3azhhyr7vh4zq0vcsfxn9v7ki-home-manager-generation.drv' failed to build
error: 1 dependencies of derivation '/nix/store/nyb4j6ylfxh3w9jxlp1c3z3irkrq47kc-unit-home-manager-perchun.service.drv' failed to build
error: 1 dependencies of derivation '/nix/store/dsvi5k2457xgfn8qxxc9h6snbcl0nvq4-system-units.drv' failed to build
error: 1 dependencies of derivation '/nix/store/xfav9m4dhxsqrdh0rj5xfhnk2mm40850-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/gzr4skxmghdfh3m5shhyb286f9xijja6-nixos-system-perchun-pc-23.11.20240314.878ef7d.drv' failed to build
```

`nix log /nix/store/05908nd888ss63axnky4xv17swjvq2jd-python3.12-pydocstyle-6.3.0.drv`: https://gist.github.com/PerchunPak/4095a5e8032ec3417db12f5d2805cf6d

### Additional context

Pydocstyle is not maintained anymore and didn't receive any updates for one year. Though isort is, and it is dependent on pydocstyle. This is why I am reporting this to isort package, and not the dead pydocstyle one.

Or should I report this to isort repository? I don't really know.

### Notify maintainers

I am not sure there are active maintainers for isort here though. (Only `couchemar` is listed in `meta.maintainers`, but they committed to isort once 6 years ago.)

### Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.81, NixOS, 23.11 (Tapir), 23.11.20240312.51063ed`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/etc/nix/path/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by PerchunPak** [comment]
For anyone who is looking for a solution: pydocstyle was deprecated in favor of [ruff](https://github.com/astral-sh/ruff), so I think migrate from isort/black/others to ruff (which is drop-in solution) is fair.

https://github.com/astral-sh/ruff#usage

TLDR:
`ruff check` - flake8
`ruff format` - black
`ruff check --select I --fix` - [isort](https://github.com/astral-sh/ruff/issues/8926)

**Comment by natsukium** [comment]
fixed in 46b2fc4e2baaed4cbc359bfcb5955f83183f87ba
